infrastructure/sentry-subgraph/src/esXai.ts

```diff
@@ -0,0 +1,87 @@
+import {  BigInt, Bytes, ethereum, log } from "@graphprotocol/graph-ts"
+import { 
+    RedemptionStarted as RedemptionStartedEvent,
+    RedemptionCancelled as RedemptionCancelledEvent,
+    RedemptionCompleted as RedemptionCompletedEvent,
+    VoucherIssued as VoucherIssuedEvent
+} from "../generated/esXai/esXai";
+import { RedemptionRequest } from "../generated/schema";
+
+export function handleRedemptionStarted(event: RedemptionStartedEvent): void {
+
+    // Create a new redemption request entity 
+    const request = new RedemptionRequest(event.params.user.toHexString() + "_"  + event.params.index.toString());
+    
+    // Decode the transaction input data to get the redemption amount and duration
+    const inputData = event.transaction.input.toHexString().slice(10); // Remove function selector
+    const decoded = ethereum.decode('(uint256,uint256)', Bytes.fromHexString(inputData) as Bytes);
+
+    if (decoded) {      
+        const decodedTuple = decoded.toTuple();
+            request.amount = decodedTuple[0].toBigInt();
+            request.duration = decodedTuple[1].toBigInt();
+    } else {
+        log.warning("Failed to decode startRedemption TX: {}", [event.transaction.hash.toHexString()]);
+        return;
+    }
```

Comment:
If we are not forced to read this from the decoded function args we should not do that. Decoding by function input has mayn risks that we should not take if we don't have to. Have you considered reading the struct from the contract ?

Reply:
We are in a subgraph event handler?

I didn't know you could read a contract struct in the subgraph event handler?

Why not just extract the function params that were called? It's done in other places in the codebase(Staking Handlers)? 

What is the alternative?

Reply:
Yes you can bind a contract instance and then call any public functions to read data. However this will be at cost of sync time as contract cals take exponentially more time when syncing.

Thats why we would not want to do this for each submission to a challenge or for each staking to any pool and we already had these events happening so there is no way to update the event if that makes sense.

This will have to be considered.

You should also have a look at the global `getInputFromEvent`, there is a good reason we use that and not decode in the handlers. This function should have enough documentation to give you an idea of the complications of decoding input data from an event.

I think we are safe to just read the redemptionRequest struct from the smart contract, if it has serious impact on syncing time of the graph we should reevaluate.

Reply:
Okay, I spent a few hours converting to read from the contract, I kept getting revert errors on read.

I couldn't seem to debug the issue and then it hit me...

Looks like the esXai contract has two redemption structs one with a postfix "Ext".

Neither are public. I kept getting a revert error when trying to read the `getRedemptionRequest` function on chain because the original version of the contract returned the old struct which requires a different function signature.

The only way to possibly make it work reading the contract would be to add the old get getRedemptionRequest back into the abi so the subgraph could call it.

Options are:
1. Stick with the currently working, deployed version using decode of function params,
2. Create a Frankenstein ABI with both function signatures and somehow try to determine in the event handler which function to use. 

My vote is #1, but I'd need to change this code back to the way it was.

Reply:
There was never a redemption created using `_redemptionRequests`, this was deployed but never enabled, then we upgraded and had to crete the new struct `_extRedemptionRequests` so we did not have to ignore hardhat storage layout protection.

Onchain there is no request saved in the "old" `_redemptionRequests`.

I cannot see why `getRedemptionRequest` would not return the `RedemptionRequestExt` for the user at the index.

Feel free to schedula a pair programming session to figure this out together.

Reply:
The original contract had a different function signature. Check the return type here:

`   function getRedemptionRequest(address account, uint256 index) public view returns (RedemptionRequest memory) {
        return _redemptionRequests[account][index];
    }`